### ISSUE TITLE
Regression: Multiline support breaks .env files with quotes in comments

### ISSUE THREAD

**Issue by BlackDex** [allan2/dotenvy#12]
Hello there,

First thanks for keeping this crate alive.
I just tried to update to the latest version, but it broke my environment.
It looks like the new multiline support also checks within comments, and that breaks everything.

It is very simple to reproduce this by modifying `dotenv/tests/common/mod.rs` to this:
```rust
pub fn make_test_dotenv() -> io::Result<TempDir> {
    tempdir_with_dotenv(
        r#"
        # Start of .env file
        # Comment line with single ' quote
        # Comment line with double " quote
        TESTKEY=test_val
        # End of .env file
        "#
    )
}
```

That will create a multiline .env file to test with. And it will break.
When  you remove the two lines having a double and single quote it will run again.

I didn't had time to look into this my self, but maybe @hoijui or @allan2 knows a quick fix for this :).

For now i will stick with the previous version of this crate.

**Comment by allan2** [comment]
Thanks for reporting this. I will look into it.

**Comment by allan2** [comment]
Hi @BlackDex,

Maybe I am misunderstanding the issue. I was unable to reproduce.

*.env*
```sh
# Start of .env file
# Comment line with single ' quote
# Comment line with double " quote
TESTKEY=test_val
# End of .env file
```

*main.rs*
```rust
use dotenvy::{dotenv, dotenv_iter, Error};

fn main() -> Result<(), Error> {
    dotenv()?;
    for item in dotenv_iter()? {
        let (key, val) = item?;
        println!("{key}={val}");  // prints `TESTKEY=test_val`
    }
    Ok(())
}
```

Could you provide more info or an example?

**Comment by BlackDex** [comment]
@allan2  If i use the example i made above and run `cargo test` it breaks for me

**Comment by BlackDex** [comment]
@allan2, i created a new branch on my fork to illustrate it.
https://github.com/allan2/dotenvy/compare/master...BlackDex:dotenvy:quote_issue

With that branch, within the `dotenv` directory both `cargo test` and `cargo run --example quote_issue` will fail

**Comment by BlackDex** [comment]
The issue here is that the new Multiline iter checks every line, also if it starts with a `#`.
Which for multiline variables could be needed of course, but it should exclude lines that start with a `#` (And any type of whitespace in front of it) when not within a `StrongOpen` i think.

Example of when it could be needed:
```bash
CODEGEN_TEST_VAR1="hello!"
CODEGEN_TEST_VAR2="'quotes within quotes'"
CODEGEN_TEST_MULTILINE1="First Line
Second Line"
CODEGEN_TEST_MULTILINE1="# First Line Comment
Second Line
#Third Line Comment
Fourth Line
"
```

**Comment by BlackDex** [comment]
@allan2 Unfortunately this is still an issue.

It even has an other regression now.
Take this for example:
```bash
# Start of .env file
# Comment line with single ' quote
# Comment line with double " quote
 # Comment line, starts with space with double " quote
 CODEGEN_TEST_VAR1="hello!"
CODEGEN_TEST_VAR2="'quotes within quotes'"
CODEGEN_TEST_VAR3="double quoted with # hash in value"
CODEGEN_TEST_VAR4='single quoted with # hash in value'
CODEGEN_TEST_VAR5=not_quoted_with_#_hash_in_value
CODEGEN_TEST_VAR5=not_quoted_with_comment_beheind #_hash_in_value
# End of .env file
```

This worked in the `0.15.1` version just fine. Now it breaks.

And the following example also breaks with the multiline feature enabled:
```bash
# Start of .env file
# Comment line with single ' quote
# Comment line with double " quote
 # Comment line, starts with space with double " quote
 CODEGEN_TEST_VAR1="hello!"
CODEGEN_TEST_VAR2="'quotes within quotes'"
CODEGEN_TEST_VAR3="double quoted with # hash in value"
CODEGEN_TEST_VAR4='single quoted with # hash in value'
CODEGEN_TEST_VAR5=not_quoted_with_#_hash_in_value
CODEGEN_TEST_VAR5=not_quoted_with_comment_beheind #_hash_in_value
CODEGEN_TEST_MULTILINE1="First Line
Second Line"
CODEGEN_TEST_MULTILINE2="# First Line Comment
Second Line
#Third Line Comment
Fourth Line
"
# End of .env file
```

If i just leave the `CODEGEN_TEST_MULTILINE1` there, and remove the rest that will work.
Now, i don't know if you want to support Lines starting with a `#` within a multiline quoted string, but besides lines starting with a string, they could be anywhere of course, which still breaks.

**Comment by LeoniePhiline** [comment]
I am also having an issue with `KEY='ABC#DEF'`:

**What's happening:**

`eval_end_state()` in the `iter` module needs me to escape the `#` despite the strong quotes `'`, as otherwise the line will be returned by the iterator as `KEY='ABC`, truncating at the `#`.

**What I tried:**

So I escaped the source line to read `KEY='ABC\#DEF'`. (Added `\`.)

But then `parse_value()` in the `parse` module dumps the `\` into the final result, as it does not expect any escaping (and never sets `escaped = true`.

So in the end, I get a value `ABC` if I do not escape, or I get `ABC\#DEF` if I do escape. Both are not the desired outcome (which is `ABC#DEF`).
